File: src/prod/src/data/utilities/ReaderWriterAsyncLock.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using ::_delete;
using namespace ktl;
using namespace Data::Utilities;


ReaderWriterAsyncLock::ReaderWriterAsyncLock()
    : waitingWritersList_(GetThisAllocator())
{
    SetConstructorStatus(waitingWritersList_.Status());
}

ReaderWriterAsyncLock::ReaderWriterAsyncLock(__in SignalOwnersTestFunctionType TestSignalOwnersAction)
    : waitingWritersList_(GetThisAllocator())
    , signalOwnersTestAction_(TestSignalOwnersAction)
{
    SetConstructorStatus(waitingWritersList_.Status());
}

ReaderWriterAsyncLock::~ReaderWriterAsyncLock()
{
}

NTSTATUS
ReaderWriterAsyncLock::Create(
    __in KAllocator& Allocator, 
    __in ULONG Tag, 
    __out ReaderWriterAsyncLock::SPtr& result)
{
    NTSTATUS status;
    ReaderWriterAsyncLock::SPtr output = _new(Tag, Allocator) ReaderWriterAsyncLock();

    if (!output)
    {
        return STATUS_INSUFFICIENT_RESOURCES;
    }

    status = output->Status();
    if (!NT_SUCCESS(status))
    {
        return status;
    }

    result = Ktl::Move(output);
    return STATUS_SUCCESS;
}

NTSTATUS
ReaderWriterAsyncLock::Create(
    __in SignalOwnersTestFunctionType TestSignalOwnersAction,
    __in KAllocator& Allocator, 
    __in ULONG Tag, 
    __out ReaderWriterAsyncLock::SPtr& result)
{
    NTSTATUS status;
    ReaderWriterAsyncLock::SPtr output = _new(Tag, Allocator) ReaderWriterAsyncLock(TestSignalOwnersAction);

    if (!output)
    {
        return STATUS_INSUFFICIENT_RESOURCES;
    }

    status = output->Status();
    if (!NT_SUCCESS(status))
    {
        return status;
    }

    result = Ktl::Move(output);
    return STATUS_SUCCESS;
}

void ReaderWriterAsyncLock::Close()
{
    KArray<KSharedPtr<AwaitableCompletionSource<bool>>> waiters(GetThisAllocator());
    K_LOCK_BLOCK(readerWriterLock_)
    {
        isClosed_ = true;

        // Set exception for all waiting writers.
        for (ULONG i = 0; i < waitingWritersList_.Count(); i++)
        {
            NTSTATUS status = waiters.Append(waitingWritersList_[i]);

            if (NT_SUCCESS(status) == false)
            {
                throw ktl::Exception(status);
            }
        }

        // Set exception for the waiting reader task.
        auto cachedWaitingReadersAcs = waitingReadersAcs_.Get();

        if (cachedWaitingReadersAcs != nullptr)
        {
            NTSTATUS status = waiters.Append(cachedWaitingReadersAcs);

            if (NT_SUCCESS(status) == false)
            {
                throw ktl::Exception(status);
            }
        }

        // Clear local state.
        waitingWritersList_.Clear();
        waitingReadersAcs_.Put(nullptr);
    }

    // Set exception for all waiters outside the lock.
    for (ULONG index = 0; index < waiters.Count(); index++)
    {
        // It is okay to synchronously complete here.
        auto completedWaiter = waiters[index];

        completedWaiter->SetException(ktl::Exception(SF_STATUS_OBJECT_CLOSED));
    }

    // Clear the collection.
    waiters.Clear();
}

ktl::Awaitable<bool> ReaderWriterAsyncLock::WaitForLock(
    __in ktl::AwaitableCompletionSource<bool> & waitingCompletionSource, 
    __in bool writer, 
    __in int timeoutInMilliSeconds)
{
    KSharedPtr<ktl::AwaitableCompletionSource<bool>> waitingAcs(&waitingCompletionSource);
    bool isLockAcquired = false;

    SharedException::CSPtr exceptionCSPtr = nullptr;

    KTimer::SPtr localTimer;
    NTSTATUS status = KTimer::Create(localTimer, GetThisAllocator(), READER_WRITER_LOCK_TAG);
    THROW_ON_FAILURE(status);

    auto timeoutTask = localTimer->StartTimerAsync(timeoutInMilliSeconds, nullptr);
    auto localWaiterTask = waitingAcs->GetAwaitable();

    co_await ktl::EitherReady(this->GetThisAllocator().GetKtlSystem(), timeoutTask, localWaiterTask);

    try
    {
        K_LOCK_BLOCK(readerWriterLock_)
        {
            if (isClosed_)
            {
                exceptionCSPtr = SharedException::Create(Ktl::Move(ktl::Exception(SF_STATUS_OBJECT_CLOSED)), GetThisAllocator());

                // This will be caught below
                auto exec = exceptionCSPtr->Info;
                throw exec;
            }

            if (waitingAcs->IsCompleted())
            {
                if (waitingAcs->IsExceptionSet())
                {
                    isLockAcquired = false;
                    // todo: use NT status codes here.
                    exceptionCSPtr = SharedException::Create(Ktl::Move(ktl::Exception(SF_STATUS_OBJECT_CLOSED)), GetThisAllocator());
                }
                else
                {
                    isLockAcquired = true;
                }
            }
            else
            {
                if (writer)
                {
                    // Timeout raced with signaling the waiting tcs (setresult).
                    // Shared ptr implements comparator that compares the raw ptr.
                    if (waitingAcs == activeWriterAcs_.Get())
                    {
                        isLockAcquired = true;
                    }
                }
                else
                {
                    // Timeout raced with signaling the waiting tcs (setresult).
                    if (waitingAcs == lastSignaledReaderAcs_)
                    {
                        isLockAcquired = true;
                    }
                }
            }

            if (isLockAcquired == false)
            {
                if (writer)
                {
                    // Dequeue waiting writer and immediately complete it w/ exception
                    waitingAcs->SetException(SF_STATUS_TIMEOUT);
                    int index = GetIndex(waitingAcs);
                    BOOLEAN wasRemoved = waitingWritersList_.Remove(index);

                    // Assert will fire if there was a delay in signaling the tcs on release writer or release reader.
                    ASSERT_IFNOT(wasRemoved == TRUE, "Delay in signaling TCS on release writer or reader");
                }
                else
                {
                    // Clean up the state that was set now that lock-acquisition has failed.
                    waitingReaderCount_--;
                    ASSERT_IFNOT(waitingReaderCount_ >= 0, "No waiting readers");
                }
            }

            if (exceptionCSPtr != nullptr)
            {
                ASSERT_IFNOT(isLockAcquired == false, "Lock should not be held processing exception during initialization");

                // This will be caught below
                auto exec = exceptionCSPtr->Info;
                throw exec;
            }
        }
    }
    catch (ktl::Exception const & e)
    {
        ASSERT_IFNOT(e.GetStatus() == SF_STATUS_OBJECT_CLOSED, "Unexpected exception. status={0}", e.GetStatus());
    }

    // Ensure the timeout task has run to completion, cancelling if necessary
    localTimer->Cancel();

    try
    {
        co_await timeoutTask;
    }
    catch (Exception const & e)
    {
        ASSERT_IFNOT(e.GetStatus() == STATUS_CANCELLED, "Unexpected exception. status={0}", e.GetStatus());
    }

    // Move the awaitable for the timed out lock to the background to be completed and cleaned up eventually
    // If the lock didn't timeout or has already run to completion, this will be a no-op
    Task timedOutLockTask = ToTask(localWaiterTask);
    ASSERT_IFNOT(timedOutLockTask.IsTaskStarted(), "Long running TCS should be awaited on in background");

    // Clean up has finished, so throw to higher layer if necessary
    if (exceptionCSPtr != nullptr)
    {
        auto exec = exceptionCSPtr->Info;
        throw exec;
    }

    co_return isLockAcquired;
}

ktl::Awaitable<bool> ReaderWriterAsyncLock::AcquireReadLockAsync(__in int timeoutInMilliSeconds)
{
    bool waitForReadLock = false;
    bool isLockAcquired = false;

    KSharedPtr<AwaitableCompletionSource<bool>> tempWaitingReadersAcs = nullptr;
    K_LOCK_BLOCK(readerWriterLock_)
    {
        if (isClosed_)
        {
            try
            {
                throw ktl::Exception(SF_STATUS_OBJECT_CLOSED);
            }
            catch (...)
            {
                throw;
            }
        }

        if (activeWriterAcs_.Get() == nullptr && waitingWritersList_.Count() == 0)
        {
            activeReaderCount_++;
            isLockAcquired = true;
        }
        else
        {
            waitingReaderCount_++;

            auto cachedWaitingReadersAcs = waitingReadersAcs_.Get();
            if (cachedWaitingReadersAcs == nullptr)
            {
                NTSTATUS status = AwaitableCompletionSource<bool>::Create(GetThisAllocator(), 0, cachedWaitingReadersAcs);

                if (NT_SUCCESS(status) == false)
                {
                    throw ktl::Exception(status);
                }

                waitingReadersAcs_.Put(Ktl::Move(cachedWaitingReadersAcs));
                cachedWaitingReadersAcs = waitingReadersAcs_.Get();
            }

            // Make a copy because await is done outside the lock.
            tempWaitingReadersAcs = cachedWaitingReadersAcs;

            // Respect timeout only if the reader needs to wait.
            waitForReadLock = true;
        }
    }

    if (waitForReadLock == false)
    {
        co_await suspend_never{};
        co_return isLockAcquired;
    }
    else
    {
        isLockAcquired = co_await WaitForLock(*tempWaitingReadersAcs, false, timeoutInMilliSeconds);
    }

    co_return isLockAcquired;
}

ktl::Awaitable<bool> ReaderWriterAsyncLock::AcquireWriteLockAsync(__in int timeoutInMilliSeconds)
{
    KSharedPtr<ktl::AwaitableCompletionSource<bool>> waitingWriterAcs = nullptr;
    NTSTATUS status = AwaitableCompletionSource<bool>::Create(GetThisAllocator(), 0, waitingWriterAcs);
    if (NT_SUCCESS(status) == false)
    {
        throw ktl::Exception(status);
    }

    bool waitForWriteLock = false;
    bool isLockAcquired = false;

    K_LOCK_BLOCK(readerWriterLock_)
    {
        if (isClosed_)
        {
            waitingWriterAcs = nullptr;
            try
            {
                throw ktl::Exception(SF_STATUS_OBJECT_CLOSED);
            }
            catch (...)
            {
                throw;
            }
        }

        if (activeWriterAcs_.Get() == nullptr && activeReaderCount_ == 0)
        {
            activeWriterAcs_.Put(Ktl::Move(waitingWriterAcs));
            waitingWriterAcs = activeWriterAcs_.Get();
            isLockAcquired = true;
        }
        else
        {
            status = waitingWritersList_.Append(waitingWriterAcs);
            if (NT_SUCCESS(status) == false)
            {
                throw ktl::Exception(status);
            }
            waitForWriteLock = true;
        }
    }

    if (waitForWriteLock == false)
    {
        co_await suspend_never{};
        co_return isLockAcquired;
    }
    else
    {
        isLockAcquired = co_await WaitForLock(*waitingWriterAcs, true, timeoutInMilliSeconds);
    }

    co_return isLockAcquired;
}

void ReaderWriterAsyncLock::ReleaseReadLock()
{
    KSharedPtr<ktl::AwaitableCompletionSource<bool>> tcsToBeSignalled = nullptr;

    K_LOCK_BLOCK(readerWriterLock_)
    {
        activeReaderCount_--;
        ASSERT_IFNOT(activeReaderCount_ >= 0, "No active readers during release lock");

        // If this is the last reader, then signal a waiting writer if one is present.
        if (activeReaderCount_ == 0 && waitingWritersList_.Count() > 0)
        {
            ASSERT_IFNOT(activeWriterAcs_.Get() == nullptr, "Null active writer TCS");
            tcsToBeSignalled = waitingWritersList_[0];
            ASSERT_IFNOT(tcsToBeSignalled != nullptr, "Null to be signalled TCS");

            activeWriterAcs_.Put(Ktl::Move(tcsToBeSignalled));
            tcsToBeSignalled = activeWriterAcs_.Get();
            BOOLEAN removed = waitingWritersList_.Remove(0);
            ASSERT_IFNOT(removed == TRUE, "Failed to remove from waitingWritersList in ReaderWriterAsyncLock::ReleaseReadLock");
        }
    }

    if (tcsToBeSignalled != nullptr)
    {
        //tcsToBeSignalled->SetResult(true);
        SignalCompletionSource(*tcsToBeSignalled);
    }
}

void ReaderWriterAsyncLock::ReleaseWriteLock()
{
    ASSERT_IFNOT(IsActiveWriter == true, "No active writers");

    KSharedPtr<ktl::AwaitableCompletionSource<bool>> tcsToBeSignalled = nullptr;

    K_LOCK_BLOCK(readerWriterLock_)
    {
        // Check for waiting writers first so that writer gets preference.
        if (waitingWritersList_.Count() > 0)
        {
            // Active writer is already true so do not have to set it, just signal the waiting writer.
            tcsToBeSignalled = waitingWritersList_[0];
            activeWriterAcs_.Put(Ktl::Move(tcsToBeSignalled));
            tcsToBeSignalled = activeWriterAcs_.Get();
            BOOLEAN removed = waitingWritersList_.Remove(0);
            ASSERT_IFNOT(removed == TRUE, "Failed to remove from waitingWritersList in ReaderWriterAsyncLock::ReleaseWriteLock");
        }
        else
        {
            auto cachedActiveWriterAcs = activeWriterAcs_.Get();
            ASSERT_IFNOT(cachedActiveWriterAcs != nullptr, "Null active writer TCS");

            activeWriterAcs_.Put(nullptr);
            if (waitingReaderCount_ > 0)
            {
                // Allow waiting readers
                tcsToBeSignalled = waitingReadersAcs_.Get();

                lastSignaledReaderAcs_ = tcsToBeSignalled;
                activeReaderCount_ = waitingReaderCount_;
                waitingReaderCount_ = 0;

                waitingReadersAcs_.Put(nullptr);
            }
        }
    }

    if (tcsToBeSignalled != nullptr)
    {
        //tcsToBeSignalled->SetResult(true);
        SignalCompletionSource(*tcsToBeSignalled);
    }
}

int ReaderWriterAsyncLock::GetIndex(__in KSharedPtr<ktl::AwaitableCompletionSource<bool>> waitingAcs)
{
    for (int i = 0; i < static_cast<int>(waitingWritersList_.Count()); i++)
    {
        if (waitingWritersList_[i] == waitingAcs)
        {
            return i;
        }
    }

    ASSERT_IF(true, "ReaderWriterAsyncLock GetIndex did not find waitingtcs");
    return -1;
}

void ReaderWriterAsyncLock::SignalCompletionSource(__in ktl::AwaitableCompletionSource<bool> & tcs)
{
    if (signalOwnersTestAction_)
    {
        signalOwnersTestAction_();
    }

    tcs.SetResult(true);
}

File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/ReaderWriterAsyncLock.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System;
    using System.Collections.Generic;
    using System.Fabric;
    using System.Fabric.Common;
    using System.Globalization;
    using System.Threading;
    using System.Threading.Tasks;

    /// <summary>
    /// Asynchronous reader writer lock.
    /// </summary>
    internal class ReaderWriterAsyncLock
    {
        /// <summary>
        /// Lock object used for mutual exclusion when readers and writers update the local state.
        /// </summary>
        private readonly object readerWriterLock = new object();

        /// <summary>
        /// Readers currently in the critical section.
        /// </summary>
        private int activeReaderCount;

        /// <summary>
        /// Indicates if there is an active writer executing the critical section or not.
        /// </summary>
        private TaskCompletionSource<bool> activeWriterTcs;

        /// <summary>
        /// Last TaskCompletionSource used the signal the waiting readers.
        /// </summary>
        private TaskCompletionSource<bool> lastSignalledReaderTcs;

        /// <summary>
        /// Indicates if the replica is closing.
        /// </summary>
        public bool IsClosed { get; private set; }

        /// <summary>
        /// Readers waiting to enter critical section.
        /// </summary>
        private int waitingReaderCount = 0;

        /// <summary>
        /// Readers wait on this tcs when there are active writers.
        /// </summary>
        private TaskCompletionSource<bool> waitingReadersTcs;

        /// <summary>
        /// Writers queue where writers wait for other writers/readers to finish.
        /// </summary>
        private List<TaskCompletionSource<bool>> waitingWritersList;

        /// <summary>
        /// Adds ability to induce failures/slow downs during hand off.
        /// </summary>
        private readonly Action signalOwnersTestAction;

        /// <summary>
        /// Initializes a new instance of the <see cref="ReaderWriterAsyncLock"/> class.
        /// </summary>
        public ReaderWriterAsyncLock()
        {
            this.activeReaderCount = 0;

            this.lastSignalledReaderTcs = null;
            this.activeWriterTcs = null;
            this.waitingReadersTcs = null;
            this.waitingWritersList = new List<TaskCompletionSource<bool>>();
            this.IsClosed = false;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="ReaderWriterAsyncLock"/> class.
        /// TEST ONLY CONSTRUCTOR - There is a reason why it is an internal constructor!
        /// </summary>
        /// <param name="signalOwnersTestAction">Used to inject failures and delays during handoff.</param>
        internal ReaderWriterAsyncLock(Action signalOwnersTestAction) : this()
        {
            Utility.Assert(signalOwnersTestAction != null, "Passed in null");
            this.signalOwnersTestAction = signalOwnersTestAction;
        }

        /// <summary>
        /// Gets the number of active readers
        /// </summary>
        /// <devnote> It is exposed for testing only.</devnote>
        public int ActiveReaderCount
        {
            get { return this.activeReaderCount; }
        }

        /// <summary>
        /// Gets the value that indicates if there is an active writer or not.
        /// </summary>
        /// <devnote> It is exposed for testing only.</devnote>
        public bool IsActiveWriter
        {
            get { return this.activeWriterTcs != null; }
        }

        /// <summary>
        /// Gets the number of waiting readers.
        /// </summary>
        /// <devnote> It is exposed for testing only.</devnote>
        public int WaitingReaderCount
        {
            get { return this.waitingReaderCount; }
        }

        /// <summary>
        /// Gets the number of waiting writers.
        /// </summary>
        /// <devnote> It is exposed for testing only.</devnote>
        public int WaitingWriterCount
        {
            get { return this.waitingWritersList.Count; }
        }

        /// <summary>
        /// Acquires read lock.
        /// </summary>
        /// <returns></returns>
        public async Task<bool> AcquireReadLockAsync(int timeoutInMilliSeconds, CancellationToken cancellationToken)
        {
            var waitForReadLock = false;
            var isLockAcquired = false;
            TaskCompletionSource<bool> tempWaitingReadersTcs = null;

            lock (this.readerWriterLock)
            {
                // Check if the replica is closing.
                if (this.IsClosed)
                {
                    throw new FabricObjectClosedException();
                }

                if (this.activeWriterTcs == null && this.waitingWritersList.Count == 0)
                {
                    this.activeReaderCount++;
                    isLockAcquired = true;
                }
                else
                {
                    this.waitingReaderCount++;

                    if (this.waitingReadersTcs == null)
                    {
                        this.waitingReadersTcs = new TaskCompletionSource<bool>();
                    }

                    // Make a copy because await is done outside the lock.
                    tempWaitingReadersTcs = this.waitingReadersTcs;

                    // Respect timeout only if the reader needs to wait.
                    waitForReadLock = true;
                }
            }

            if (!waitForReadLock)
            {
                return isLockAcquired;
            }
            else
            {
                isLockAcquired = await this.WaitForLock(tempWaitingReadersTcs, false, timeoutInMilliSeconds).ConfigureAwait(false);
            }

            return isLockAcquired;
        }

        /// <summary>
        /// Acquires write lock.
        /// </summary>
        /// <returns></returns>
        public async Task<bool> AcquireWriteLockAsync(int timeoutInMilliSeconds, CancellationToken cancellationToken)
        {
            TaskCompletionSource<bool> waitingWriterTcs = new TaskCompletionSource<bool>();
            var waitForWriteLock = false;
            var isLockAcquired = false;

            lock (this.readerWriterLock)
            {
                // check if the replica is closing.
                if (this.IsClosed)
                {
                    throw new FabricObjectClosedException();
                }

                if (this.activeWriterTcs == null && this.activeReaderCount == 0)
                {
                    this.activeWriterTcs = waitingWriterTcs;
                    isLockAcquired = true;
                }
                else
                {
                    // Queue a waiting writer task.
                    this.waitingWritersList.Add(waitingWriterTcs);
                    waitForWriteLock = true;
                }
            }

            if (!waitForWriteLock)
            {
                return isLockAcquired;
            }
            else
            {
                isLockAcquired = await this.WaitForLock(waitingWriterTcs, true, timeoutInMilliSeconds).ConfigureAwait(false);
            }

            return isLockAcquired;
        }

        /// <summary>
        /// Does clean up on close.
        /// </summary>
        public void Close()
        {
            var waiters = new List<TaskCompletionSource<bool>>();
            lock (this.readerWriterLock)
            {
                this.IsClosed = true;

                // Set exception for all waiting writers.
                waiters.AddRange(this.waitingWritersList);

                // Set exception for the waiting reader task.
                if (this.waitingReadersTcs != null)
                {
                    waiters.Add(this.waitingReadersTcs);
                }

                // Clear local state.
                this.waitingWritersList.Clear();
                this.waitingReadersTcs = null;
            }

            // Set exception for all waiters outside the lock.
            for (var index = 0; index < waiters.Count; index++)
            {
                var completedWaiter = waiters[index];
                this.ExecuteTaskWithPreferFairness(
                    () =>
                    {
                        completedWaiter.Task.IgnoreExceptionVoid();
                        completedWaiter.TrySetException(new FabricObjectClosedException());
                    });
            }

            // Clear the collection.
            waiters.Clear();
        }

        /// <summary>
        /// Releases read lock.
        /// </summary>
        public void ReleaseReadLock()
        {
            TaskCompletionSource<bool> tcsToBeSignalled = null;
            lock (this.readerWriterLock)
            {
                this.activeReaderCount--;
                Utility.Assert(
                    this.activeReaderCount >= 0,
                    "user is trying to realease a lock that has not been acquired");

                // If this is the last reader, then signal a waiting writer if one is present.
                if (this.activeReaderCount == 0 && this.waitingWritersList.Count > 0)
                {
                    Utility.Assert(this.activeWriterTcs == null, "Current active writer cannot be null");
                    tcsToBeSignalled = this.waitingWritersList[0];
                    Utility.Assert(tcsToBeSignalled != null, "Writer to be signaled cannot be null");

                    this.activeWriterTcs = tcsToBeSignalled;
                    this.waitingWritersList.Remove(tcsToBeSignalled);
                }
            }

            // Set result outside the lock.
            if (tcsToBeSignalled != null)
            {
                this.ExecuteTaskWithPreferFairness(
                    () =>
                    {
                        this.InvokeSignalOwnersActionIfNecessary();
                        tcsToBeSignalled.SetResult(true);
                    });
            }
        }

        /// <summary>
        /// Realeases write lock.
        /// </summary>
        public void ReleaseWriteLock()
        {
            Utility.Assert(this.IsActiveWriter, "Cannot release a lock that was not acquired.");

            TaskCompletionSource<bool> tcsToBeSignalled = null;
            lock (this.readerWriterLock)
            {
                // Check for waiting writers first so that writer gets preference.
                if (this.waitingWritersList.Count > 0)
                {
                    // Active writer is already true so do not have to set it, just signal the waiting writer.
                    tcsToBeSignalled = this.waitingWritersList[0];
                    this.activeWriterTcs = tcsToBeSignalled;
                    this.waitingWritersList.Remove(tcsToBeSignalled);
                }
                else
                {
                    // Active writer has finished and it is the reader's turn.
                    Utility.Assert(this.activeWriterTcs != null, "active writer tcs should not be null");
                    this.activeWriterTcs = null;
                    if (this.waitingReaderCount > 0)
                    {
                        // Allow waiting readers
                        tcsToBeSignalled = this.waitingReadersTcs;

                        this.lastSignalledReaderTcs = this.waitingReadersTcs;
                        this.activeReaderCount = this.waitingReaderCount;
                        this.waitingReaderCount = 0;

                        this.waitingReadersTcs = null;
                    }
                }
            }

            // Hand off the lock the new writer owner or read owners.
            if (tcsToBeSignalled != null)
            {
                this.ExecuteTaskWithPreferFairness(
                    () =>
                        {
                            this.InvokeSignalOwnersActionIfNecessary();
                            tcsToBeSignalled.SetResult(true);
                        });
            }
        }

        /// <summary>
        /// Waits for the given TCS to finish within the timeout specified.
        /// </summary>
        private async Task<bool> WaitForLock(
            TaskCompletionSource<bool> waitingTcs,
            bool writer,
            int timeoutInMilliSeconds)
        {
            // Respect timeout.
            var shouldTimeoutTaskBeCancelled = false;
            var isLockAcquired = false;
            Exception exception = null;

            using (var cancellationTokenSource = new CancellationTokenSource())
            {
                var timeoutTask = Task.Delay(timeoutInMilliSeconds, cancellationTokenSource.Token);

                await Task.WhenAny(timeoutTask, waitingTcs.Task).ConfigureAwait(false);

                lock (this.readerWriterLock)
                {
                    if (this.IsClosed)
                    {
                        throw new FabricObjectClosedException();
                    }

                    if (waitingTcs.Task.IsCompleted)
                    {
                        shouldTimeoutTaskBeCancelled = true;
                        if (waitingTcs.Task.Status == TaskStatus.RanToCompletion)
                        {
                            // Successful completion.
                            isLockAcquired = true;
                        }
                        else if (waitingTcs.Task.IsFaulted)
                        {
                            // Set lock acquired status to false.
                            isLockAcquired = false;
                            exception = waitingTcs.Task.Exception.InnerException ?? waitingTcs.Task.Exception;
                        }
                        else if (waitingTcs.Task.IsCanceled)
                        {
                            // Set lock acquired status to false.
                            isLockAcquired = false;
                            exception = new OperationCanceledException();
                        }
                    }
                    else
                    {
                        if (writer)
                        {
                            // Timeout raced with signaling the waiting tcs (setresult).
                            isLockAcquired = (waitingTcs == this.activeWriterTcs);
                        }
                        else
                        {
                            // Timeout raced with signaling the waiting tcs (setresult).
                            isLockAcquired = (waitingTcs == this.lastSignalledReaderTcs);
                            Utility.Assert(!isLockAcquired || this.activeReaderCount > 0, "If lock was acquired through race, active reader count must be greater than 0.");
                        }
                    }

                    if (false == isLockAcquired)
                    {
                        if (writer)
                        {
                            // Dequeue waiting writer.
                            var wasRemoved = this.waitingWritersList.Remove(waitingTcs);

                            // Assert will fire if there was a delay in signaling ths tcs on release writer or release reader.
                            Utility.Assert(wasRemoved, "failed to remove writer tcs");
                        }
                        else
                        {
                            // Clean up the state that was set now that lock-acquisition has failed.
                            this.waitingReaderCount--;

                            // Assert will fire if there was a delay in signaling ths tcs on release writer until bug 6720586 is fixed.
                            Utility.Assert(this.waitingReaderCount >= 0, "waiting reader count {0} cannot be lesser than zero", this.waitingReaderCount);
                        }
                    }
                }

                // This should be done outside the lock.
                if (shouldTimeoutTaskBeCancelled)
                {
                    // Cancel the tiemout task.
                    cancellationTokenSource.Cancel();

                    try
                    {
                        // Await the task and swallow the exception explicitly if the task has not yet completed since the task has been cancelled,
                        // else await would return right away.
                        await timeoutTask.ConfigureAwait(false);
                    }
                    catch (TaskCanceledException)
                    {
                        // Swallow TaskCanceledException.
                    }
                }

                if (null != exception)
                {
                    Utility.Assert(false == isLockAcquired, "lock acquisition status shoud be false");
                    throw exception;
                }
            }

            return isLockAcquired;
        }

        /// <summary>
        /// Executes the given action with prefer fairness option.
        /// </summary>
        /// <param name="action">action to be exceuted.</param>
        /// <devnote>This needs to move into utility and needs more overloads.</devnote>
        private void ExecuteTaskWithPreferFairness(Action action)
        {
            Task.Factory.StartNew(action, TaskCreationOptions.PreferFairness).IgnoreExceptionVoid();
        }

        private void InvokeSignalOwnersActionIfNecessary()
        {
            Action snapHandOffTestAction = this.signalOwnersTestAction;
            if (snapHandOffTestAction == null)
            {
                return;
            }

            // TODO: convert to structured tracing and add the correct trace type
            AppTrace.TraceSource.WriteWarning(string.Format(CultureInfo.InvariantCulture, "RWLock"), "invoking test hook");
            snapHandOffTestAction.Invoke();
        }
    }
}